### ISSUE TITLE
Creating a collection with an empty name corrupts data

### ISSUE THREAD

**Issue by alexsegura** [typesense/typesense#251]
## Description

When creating a collection with an empty name, the server crashes and won't restart unless you remove the data. 

## Steps to reproduce

Using the PHP client: 

```php
$schema = [
    'name' => null,
    'fields' => [
        ['name' => 'name', 'type' => 'string'],
        ['name' => 'created_at', 'type' => 'int32'],
    ],
    'default_sorting_field' => 'created_at'
];
$client->collections->create($schema);
```

## Expected Behavior

The API should return HTTP 400 with a message saying:

> Cannot create a collection with an empty name

## Actual Behavior

The server crashes: 

```
typesense_1     | I20210403 01:06:33.688689     1 typesense_server_utils.cpp:301] Starting Typesense 0.19.0
typesense_1     | I20210403 01:06:33.690609     1 typesense_server_utils.cpp:304] Typesense is using jemalloc.
typesense_1     | I20210403 01:06:33.730222     1 typesense_server_utils.cpp:405] Starting API service...
typesense_1     | I20210403 01:06:33.730813    82 typesense_server_utils.cpp:210] Since no --nodes argument is provided, starting a single node Typesense cluster.
typesense_1     | I20210403 01:06:33.734120     1 http_server.cpp:189] Typesense has started listening on port 8108
typesense_1     | I20210403 01:06:33.744827    82 server.cpp:1045] Server[braft::RaftStatImpl+braft::FileServiceImpl+braft::RaftServiceImpl+braft::CliServiceImpl] is serving on port=8107.
typesense_1     | I20210403 01:06:33.744860    82 server.cpp:1048] Check out http://1c913ea63cb1:8107 in web browser.
typesense_1     | I20210403 01:06:33.747742    82 log.cpp:674] Use crc32c as the checksum type of appending entries
typesense_1     | I20210403 01:06:33.747857    82 log.cpp:1158] log load_meta /data/state/log/log_meta first_log_index: 1 time: 56
typesense_1     | I20210403 01:06:33.748559    82 log.cpp:1098] load open segment, path: /data/state/log first_index: 1
typesense_1     | I20210403 01:06:33.748966    82 raft_meta.cpp:521] Loaded single stable meta, path /data/state/meta term 3 votedfor 172.18.0.23:8107:8108 time: 193
typesense_1     | I20210403 01:06:33.750368    88 raft_server.cpp:401] on_snapshot_load
typesense_1     | I20210403 01:06:33.755148    88 store.h:281] rm /data/db success
typesense_1     | I20210403 01:06:33.758131    88 store.h:291] copy snapshot /data/state/snapshot/snapshot_00000000000000000004/db_snapshot to /data/db success
typesense_1     | I20210403 01:06:33.778117    88 store.h:305] DB open success!
typesense_1     | I20210403 01:06:33.779026    88 raft_server.cpp:384] Loading collections from disk...
typesense_1     | I20210403 01:06:33.779731    88 collection_manager.cpp:71] CollectionManager::load()
typesense_1     | I20210403 01:06:33.781149    88 auth_manager.cpp:30] Indexing 0 API key(s) found on disk.
typesense_1     | I20210403 01:06:33.781222    88 collection_manager.cpp:94] Found 0 collection(s) on disk.
typesense_1     | I20210403 01:06:33.781286    88 raft_server.cpp:389] Finished loading collections from disk.
typesense_1     | I20210403 01:06:33.781344    88 raft_server.h:245] Configuration of this group is 172.18.0.23:8107:8108
typesense_1     | I20210403 01:06:33.781433    88 snapshot_executor.cpp:264] node default_group:172.18.0.23:8107:8108 snapshot_load_done, last_included_index: 4 last_included_term: 2 peers: "172.18.0.23:8107:8108"
typesense_1     | I20210403 01:06:33.781675    82 raft_meta.cpp:521] Loaded single stable meta, path /data/state/meta term 3 votedfor 172.18.0.23:8107:8108 time: 31
typesense_1     | I20210403 01:06:33.781697    82 node.cpp:576] node default_group:172.18.0.23:8107:8108 init, term: 3 last_log_id: (index=8,term=3) conf: 172.18.0.23:8107:8108 old_conf:
typesense_1     | I20210403 01:06:33.782559    82 node.cpp:1549] node default_group:172.18.0.23:8107:8108 term 3 start vote and grant vote self
typesense_1     | I20210403 01:06:33.792598    82 raft_meta.cpp:546] Saved single stable meta, path /data/state/meta term 4 votedfor 172.18.0.23:8107:8108 time: 8272
typesense_1     | I20210403 01:06:33.792671    82 node.cpp:1783] node default_group:172.18.0.23:8107:8108 term 4 become leader of group 172.18.0.23:8107:8108
typesense_1     | I20210403 01:06:33.792860    82 typesense_server_utils.cpp:261] Typesense peering service is running on 172.18.0.23:8107
typesense_1     | I20210403 01:06:33.792886    82 typesense_server_utils.cpp:262] Snapshot interval configured as: 3600s
typesense_1     | I20210403 01:06:33.792919    82 raft_server.cpp:430] Term: 4, last_index index: 8, committed_index: 0, known_applied_index: 4, applying_index: 0, pending_index: 9, disk_index: 8, pending_queue_size: 1
typesense_1     | W20210403 01:06:33.793000    82 node.cpp:811] [default_group:172.18.0.23:8107:8108 ] Refusing concurrent configuration changing
typesense_1     | E20210403 01:06:33.793161    88 raft_server.h:57] Peer refresh failed, error: Doing another configuration change
typesense_1     | terminate called after throwing an instance of 'nlohmann::detail::type_error'
typesense_1     |   what():  [json.exception.type_error.302] type must be string, but is null
typesense_1     | E20210403 01:06:34.100122     1 backward.hpp:4196] Stack trace (most recent call last):
typesense_1     | E20210403 01:06:34.100176     1 backward.hpp:4196] #18   Object "/opt/typesense-server", at 0x86ab30, in _start
typesense_1     | E20210403 01:06:34.100191     1 backward.hpp:4196] #17   Object "/lib/x86_64-linux-gnu/libc-2.23.so", at 0x7f2efee7a83f, in __libc_start_main
typesense_1     | E20210403 01:06:34.100204     1 backward.hpp:4196] #16   Source "/typesense/src/main/typesense_server.cpp", line 124, in main [0x7b1d1f]
typesense_1     | E20210403 01:06:34.100211     1 backward.hpp:4196] #15   Source "/typesense/src/typesense_server_utils.cpp", line 408, in run_server [0x98ae92]
typesense_1     | E20210403 01:06:34.100221     1 backward.hpp:4196] #14   Source "/typesense/src/http_server.cpp", line 195, in run [0x92f720]
typesense_1     | E20210403 01:06:34.100230     1 backward.hpp:4196] #13   Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/socket/evloop.c.h", line 619, in h2o_evloop_run [0x9aa0d4]
typesense_1     | E20210403 01:06:34.100240     1 backward.hpp:4196] #12   Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/socket/evloop.c.h", line 567, in run_pending [0x9a92e5]
typesense_1     | E20210403 01:06:34.100253     1 backward.hpp:4196] #11 | Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/socket/evloop.c.h", line 534, in read_on_ready
typesense_1     | E20210403 01:06:34.100270     1 backward.hpp:4196]       Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/socket/evloop.c.h", line 248, in run_socket [0x9a91f8]
typesense_1     | E20210403 01:06:34.100279     1 backward.hpp:4196] #10 | Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/multithread.c", line 90, in queue_cb
typesense_1     | E20210403 01:06:34.100293     1 backward.hpp:4196]       Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/multithread.c", line 64, in h2o_buffer_consume [0x9a782d]
typesense_1     | E20210403 01:06:34.100306     1 backward.hpp:4196] #9    Source "/typesense/include/http_data.h", line 412, in on_message [0x9347eb]
typesense_1     | E20210403 01:06:34.100343     1 backward.hpp:4196] #8    Source "/typesense/src/core_api.cpp", line 1115, in raft_write_send_response [0x916899]
typesense_1     | E20210403 01:06:34.100360     1 backward.hpp:4196] #7  | Source "/typesense/src/core_api.cpp", line 136, in operator std::__cxx11::basic_string<char><std::__cxx11::basic_string<char> >
typesense_1     | E20210403 01:06:34.100486     1 backward.hpp:4196]       Source "/typesense/include/json.hpp", line 18999, in post_create_collection [0x91fd66]
typesense_1     | E20210403 01:06:34.100512     1 backward.hpp:4196] #6  | Source "/typesense/include/json.hpp", line 18709, in from_json<const nlohmann::basic_json<>&, std::__cxx11::basic_string<char> >
typesense_1     | E20210403 01:06:34.100533     1 backward.hpp:4196]     | Source "/typesense/include/json.hpp", line 4203, in operator()<nlohmann::basic_json<>, std::__cxx11::basic_string<char> >
typesense_1     | E20210403 01:06:34.100554     1 backward.hpp:4196]     | Source "/typesense/include/json.hpp", line 3608, in from_json<nlohmann::basic_json<> >
typesense_1     | E20210403 01:06:34.100780     1 backward.hpp:4196]       Source "/typesense/include/json.hpp", line 3301, in get<std::__cxx11::basic_string<char> > [0x8c8fee]
typesense_1     | E20210403 01:06:34.100787     1 backward.hpp:4196] #5    Source "../../../../libstdc++-v3/libsupc++/eh_throw.cc", line 95, in __cxa_throw [0x12ecde3]
typesense_1     | E20210403 01:06:34.100811     1 backward.hpp:4196] #4    Source "../../../../libstdc++-v3/libsupc++/eh_terminate.cc", line 58, in terminate [0x12ecc90]
typesense_1     | E20210403 01:06:34.100993     1 backward.hpp:4196] #3    Source "../../../../libstdc++-v3/libsupc++/eh_terminate.cc", line 48, in abort [0x12ecc25]
typesense_1     | E20210403 01:06:34.101012     1 backward.hpp:4196] #2    Source "../../../../libstdc++-v3/libsupc++/vterminate.cc", line 95, in __verbose_terminate_handler [0x7a06da]
typesense_1     | E20210403 01:06:34.101027     1 backward.hpp:4196] #1    Object "/lib/x86_64-linux-gnu/libc-2.23.so", at 0x7f2efee91039, in abort
typesense_1     | E20210403 01:06:34.101044     1 backward.hpp:4196] #0    Object "/lib/x86_64-linux-gnu/libc-2.23.so", at 0x7f2efee8f438, in raise
typesense_1     | Aborted (Signal sent by tkill() 1 0)
```

## Metadata

**Typesense Version**: 0.19

**OS**: MacOSX/Docker

**Comment by cboothe** [comment]
This has happened to me too!

**Comment by kishorenc** [comment]
Yikes, will add validation for this.

**Comment by kishorenc** [comment]
Fixed in `typesense/typesense:0.20.0.rc28` Docker image. Can you please confirm?

**Comment by kishorenc** [comment]
Addressed as part of `v0.20.0` released last week.
